src/http.rs

```diff
@@ -1,105 +1,72 @@
 use anyhow::{anyhow, Result};
-use compositor_common::{
-    scene::{InputId, OutputId, Resolution, SceneSpec},
-    transformation::{TransformationRegistryKey, TransformationSpec},
-    SpecValidationError,
-};
+use compositor_common::SpecValidationError;
+use compositor_pipeline::pipeline;
 use compositor_render::registry;
 use log::{error, info};
-use serde::{Deserialize, Serialize};
+
 use serde_json::json;
-use std::{error::Error, io::Cursor, net::SocketAddr, sync::Arc, thread};
+use std::{error::Error, io::Cursor, net::SocketAddr, thread};
 use tiny_http::{Header, Response, StatusCode};
 
-use crate::rtp_sender::EncoderSettings;
-
-use super::state::State;
-
-#[derive(Serialize, Deserialize)]
-pub struct RegisterInputRequest {
-    pub id: InputId,
-    pub port: u16,
-}
-
-#[derive(Serialize, Deserialize)]
-pub struct RegisterOutputRequest {
-    pub id: OutputId,
-    pub port: u16,
-    pub ip: String,
-    pub resolution: Resolution,
-    pub encoder_settings: EncoderSettings,
-}
-
-#[derive(Serialize, Deserialize)]
-#[serde(tag = "type", rename_all = "snake_case")]
-enum Request {
-    RegisterInput(RegisterInputRequest),
-    RegisterOutput(RegisterOutputRequest),
-    UpdateScene(SceneSpec),
-    RegisterTransformation {
-        key: TransformationRegistryKey,
-        transform: TransformationSpec,
-    },
-    Init,
-    Start,
-}
+use crate::api::{Api, Request};
 
 pub struct Server {
     server: tiny_http::Server,
-    state: Arc<State>,
     content_type_json: Header,
 }
 
 impl Server {
-    pub fn new(port: u16, state: Arc<State>) -> Self {
+    pub fn new(port: u16) -> Self {
         Self {
             server: tiny_http::Server::http(SocketAddr::from(([0, 0, 0, 0], port))).unwrap(),
-            state,
             content_type_json: Header::from_bytes(&b"Content-Type"[..], &b"application/json"[..])
                 .unwrap(),
         }
     }
 
     pub fn start(self) {
         info!("Listening on port {}", self.server.server_addr());
+        let mut api = None;
         for mut raw_request in self.server.incoming_requests() {
-            let result = self.handle_request_before_init(&mut raw_request);
-            let should_abort = result.is_ok();
-            self.send_response(raw_request, result);
-            if should_abort {
-                break;
+            let result = self
+                .handle_request_before_init(&mut raw_request)
+                .and_then(Api::new);
+            match result {
+                Ok(new_api) => {
+                    self.send_response(raw_request, Ok(()));
+                    api = Some(new_api);
+                    break;
+                }
+                Err(err) => {
+                    self.send_response(raw_request, Err(err));
+                }
```

Comment:
If you move this logic to separate function (sth like `handle_init_request(..) -> Api`) returning `Api`, then you can remove this `Option` type and unwrap